pkg/ingester/ingester.go

```diff
@@ -2893,33 +2956,34 @@ func (i *Ingester) shipBlocks(ctx context.Context, allowed *util.AllowedTenants)
 	})
 }
 
-func (i *Ingester) compactionLoop(ctx context.Context) error {
+// compactionServiceRunning is the running function of internal service responsible to periodically
+// compact TSDB Head.
+func (i *Ingester) compactionServiceRunning(ctx context.Context) error {
 	// At ingester startup, spread the first compaction over the configured compaction
 	// interval. Then, the next compactions will happen at a regular interval. This logic
 	// helps to have different ingesters running the compaction at a different time,
 	// effectively spreading the compactions over the configured interval.
-	firstHeadCompactionInterval := i.cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval
-	if i.cfg.BlocksStorageConfig.TSDB.HeadCompactionIntervalJitterEnabled {
-		firstHeadCompactionInterval = util.DurationWithNegativeJitter(i.cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval, 1)
-	}
-
-	ticker := time.NewTicker(firstHeadCompactionInterval)
-	defer ticker.Stop()
-	tickerRunOnce := false
+	firstInterval, standardInterval := i.compactionServiceInterval()
+	stopTicker, tickerChan := util.NewVariableTicker(firstInterval, standardInterval)
+	defer stopTicker()
 
 	for ctx.Err() == nil {
 		select {
-		case <-ticker.C:
+		case <-tickerChan:
 			// The forcedCompactionMaxTime has no meaning because force=false.
 			i.compactBlocks(ctx, false, 0, nil)
 
 			// Check if any TSDB Head should be compacted to reduce the number of in-memory series.
 			i.compactBlocksToReduceInMemorySeries(ctx, time.Now())
 
-			// Run it at a regular (configured) interval after the first compaction.
-			if !tickerRunOnce {
-				ticker.Reset(i.cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval)
-				tickerRunOnce = true
+			// Check if the desired interval has changed. We only compare the standard interval
+			// before the first interval may be random due to jittering.
+			if newFirstInterval, newStandardInterval := i.compactionServiceInterval(); standardInterval != newStandardInterval {
+				// Stop the previous ticker before creating a new one.
+				stopTicker()
+
+				standardInterval = newStandardInterval
+				stopTicker, tickerChan = util.NewVariableTicker(newFirstInterval, newStandardInterval)
 			}
 
 		case req := <-i.forceCompactTrigger:
```

Comment:
Shall we enable force-compact before Ingester is Running? They switch state of DB and block push requests, which may be interfering with ingest-reader catch-up. But we could also easily block that in the flush handler (not allow flushes if ingester is not Running yet).

Reply:
Good call. I think it's something do regardless of ingester storage. Done in https://github.com/grafana/mimir/pull/7486/commits/065da2fed67618de48a88c4112117ec8e9695f12